src/Processors/Transforms/Streaming/GlobalAggregatingTransformWithSubstream.cpp

```diff
@@ -16,50 +23,90 @@ GlobalAggregatingTransformWithSubstream::GlobalAggregatingTransformWithSubstream
+             /// Field deserializer
+             [this](std::any & field, ReadBuffer & rb) {
+                 auto & data = std::any_cast<RetractedDataVariantsPtr &>(field);
+                 params->aggregator.recover(*data, rb);
+             }});
+    }
+    return substream_ctx;
+}
+
+std::pair<bool, bool>
+GlobalAggregatingTransformWithSubstream::executeOrMergeColumns(Chunk & chunk, const SubstreamContextPtr & substream_ctx)
+{
+    if (params->emit_changelog)
+    {
+        assert(!params->only_merge);
+
+        auto num_rows = chunk.getNumRows();
+        auto & retracted_variants = substream_ctx->getField<RetractedDataVariantsPtr>();
+        auto & aggregated_variants = substream_ctx->variants;
+
+        return params->aggregator.executeAndRetractOnBlock(
+            chunk.detachColumns(), 0, num_rows, aggregated_variants, *retracted_variants, key_columns, aggregate_columns, no_more_keys);
+    }
+    else
+        return AggregatingTransformWithSubstream::executeOrMergeColumns(chunk, substream_ctx);
+}
+
 /// Finalize what we have in memory and produce a finalized Block
 /// and push the block to downstream pipe
 void GlobalAggregatingTransformWithSubstream::finalize(const SubstreamContextPtr & substream_ctx, const ChunkContextPtr & chunk_ctx)
 {
     assert(substream_ctx);
 
+    SCOPE_EXIT({
+        substream_ctx->resetRowCounts();
+        substream_ctx->finalized_watermark = chunk_ctx->getWatermark();
+    });
+
     /// If there is no new data, don't emit aggr result
     if (!substream_ctx->hasNewData())
         return;
 
-    auto start = MonotonicMilliseconds::now();
-    doFinalize(substream_ctx, chunk_ctx);
-    auto end = MonotonicMilliseconds::now();
-
-    LOG_INFO(log, "Took {} milliseconds to finalize aggregation", end - start);
-}
-
-void GlobalAggregatingTransformWithSubstream::doFinalize(const SubstreamContextPtr & substream_ctx, const ChunkContextPtr & chunk_ctx)
-{
-    SCOPE_EXIT({ substream_ctx->resetRowCounts(); });
-
     auto & variants = substream_ctx->variants;
     if (variants.empty())
         return;
 
-    Block block;
-    if (params->final)
-    {
-        auto results = params->aggregator.convertToBlocksFinal(variants, ConvertAction::STREAMING_EMIT, 1);
-        assert(results.size() == 1);
-        block = std::move(results.back());
+    assert(variants.type != AggregatedDataVariants::Type::without_key);
 
-        if (params->emit_version)
-            emitVersion(block, substream_ctx);
+    if (unlikely(variants.isTwoLevel()))
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Two level merge is not implemented in global aggregation");
+
+    auto start = MonotonicMilliseconds::now();
+    if (params->emit_changelog)
+    {
+        auto [retracted_chunk, chunk]
+            = AggregatingHelper::convertSingleLevelToChangelog(variants, *substream_ctx->getField<RetractedDataVariantsPtr>(), *params);
+        setCurrentChunk(std::move(chunk), chunk_ctx, std::move(retracted_chunk));
```

Comment:
If emit changlog, it cannot emit version at the same time?

Reply:
Yes, 'emit_version()' is not supported in global aggregation emit changelog,  it has been checked in ctor.